hudi-common/src/main/java/org/apache/hudi/common/util/ClusteringUtils.java

```diff
@@ -70,11 +70,67 @@ public class ClusteringUtils {
   public static Stream<Pair<HoodieInstant, HoodieClusteringPlan>> getAllPendingClusteringPlans(
       HoodieTableMetaClient metaClient) {
     List<HoodieInstant> pendingClusterInstants =
-        metaClient.getActiveTimeline().filterPendingClusterTimeline().getInstants();
+        metaClient.getActiveTimeline().filterPendingReplaceOrClusteringTimeline().getInstants();
     return pendingClusterInstants.stream().map(instant -> getClusteringPlan(metaClient, instant))
         .filter(Option::isPresent).map(Option::get);
   }
 
+  /**
+   * Returns the pending clustering instant. This can be older pending replace commit or a new
+   * clustering inflight commit. After HUDI-7905, all the requested and inflight clustering instants
+   * use clustering action instead of replacecommit.
+   */
+  public static Option<HoodieInstant> getInflightClusteringInstant(String timestamp, HoodieActiveTimeline activeTimeline) {
+    HoodieTimeline pendingReplaceOrClusterTimeline = activeTimeline.filterPendingReplaceOrClusteringTimeline();
+    HoodieInstant inflightInstant = HoodieTimeline.getClusteringCommitInflightInstant(timestamp);
+    if (pendingReplaceOrClusterTimeline.containsInstant(inflightInstant)) {
+      return Option.of(inflightInstant);
+    }
+    inflightInstant = HoodieTimeline.getReplaceCommitInflightInstant(timestamp);
+    return Option.ofNullable(pendingReplaceOrClusterTimeline.containsInstant(inflightInstant) ? inflightInstant : null);
+  }
+
+  /**
+   * Returns the pending clustering instant. This can be older requested replace commit or a new
+   * clustering requested commit. After HUDI-7905, all the requested and inflight clustering instants
+   * use clustering action instead of replacecommit.
+   */
+  public static Option<HoodieInstant> getRequestedClusteringInstant(String timestamp, HoodieActiveTimeline activeTimeline) {
+    HoodieTimeline pendingReplaceOrClusterTimeline = activeTimeline.filterPendingReplaceOrClusteringTimeline();
+    HoodieInstant requestedInstant = HoodieTimeline.getClusteringCommitRequestedInstant(timestamp);
+    if (pendingReplaceOrClusterTimeline.containsInstant(requestedInstant)) {
+      return Option.of(requestedInstant);
+    }
+    requestedInstant = HoodieTimeline.getReplaceCommitRequestedInstant(timestamp);
+    return Option.ofNullable(pendingReplaceOrClusterTimeline.containsInstant(requestedInstant) ? requestedInstant : null);
+  }
+
+  /**
+   * Transitions the provided clustering instant fron inflight to complete based on the clustering
+   * action type. After HUDI-7905, the new clustering commits are written with clustering action.
+   */
+  public static void transitionClusterInflightToComplete(boolean shouldLock, HoodieInstant clusteringInstant,
```

Comment:
oh 1 thing. if there is a pending replacecommit(just requested) in timeline just before upgrade, then we need to use replacecommit (and not clustering) for inflight right?
are we handling this case.